src/Perpetuum/Services/EventServices/EventProcessors/EnvironmentalEffectHandler.cs

```diff
@@ -59,47 +78,14 @@ private void OnStateChange()
                 return;
 
             ZoneEffect nextEffect = null;
-            if (_gameTime.IsDay)
-            {
-                if (_weatherState.IsGoodWeather)
-                {
-                    nextEffect = GetEffect(EffectType.effect_day_clear);
-                }
-                else if (_weatherState.IsBadWeather)
-                {
-                    nextEffect = GetEffect(EffectType.effect_day_overcast);
-                }
-                else
-                {
-                    nextEffect = GetEffect(EffectType.effect_day);
-                }
-            }
-            else if (_gameTime.IsNight)
-            {
-                if (_weatherState.IsGoodWeather)
-                {
-                    nextEffect = GetEffect(EffectType.effect_night_clear);
-                }
-                else if (_weatherState.IsBadWeather)
-                {
-                    nextEffect = GetEffect(EffectType.effect_night_overcast);
-                }
-                else
-                {
-                    nextEffect = GetEffect(EffectType.effect_night);
-                }
-            }
-            else
+
+            //In cases where DayState is supposed to stay "neutral", WeatherState cannot also be "neutral" if we are to follow the logic this change is replacing.
```

Comment:
This can happen when both

 **WeatherInfo**(Between GOOD_WEATHER - 100 and BAD_WEATHER 200)
 and
 **GameTimeInfo** (Between NIGHT_END - 100 and DAY_START 200) 
 
 hits some of their undefined gaps at the same time, where it's neither Day or Night / Good or Bad.
 
Old logic just did nothing in that case, but here we would try to lookup a value in the _weatherDict which does not exist (neutral, neutral).

So to avoid that lookup, I just return;
But I'm not sure if we ever have a scenario where both of them hits these gaps at the same time.

Reply:
Yes, both weather and daytime can have a neutral state and not have any zone effect, so make sure this is still allowed with your lookup

Reply:
Thank you for the insight!

I have now flipped the logic of the if statement, so that if both are in the neutral state we don't touch the initial `ZoneEffect nextEffect = null` and pass it to the `isSameEffect` unchanged. It should have the same result as you described.

If you instead prefer that there is a null entry in the dictionary, I'd have to look into how to solve that.

Reply:
I think you can just add another entry with `{(neutral, neutral): null}` and let it look up and get the null value from the dict.  
That way you avoid that one remaining if statement wrapping the lookup - that way the mapping is all contained in the dictionary.

Reply:
I've pushed [#a35a0f2](https://github.com/OpenPerpetuum/PerpetuumServer/pull/278/commits/a35a0f2b15e2a152957c3ba86680f7db8ec996b4) which tries to achieve that.
There is no longer any if statement, and the dictionary contains an entry for neutral+neutral which returns null.

The alternative to having `dynamic` instead of `EffectType` as the expected values in the dictionary, was to mark it with `EffectType?` as a nullable. But that had some ripple-effects throughout the class, affecting the other dictionary and methods. 

So I think this looks a little cleaner in comparison.

Reply:
Actually, hang on a sec.  I just realized you are creating a dictionary of EffectTypes and not the ZoneEffect instances - which means we can simplify this into one dictionary!  (woo!) No dynamic type, and you have a nullable class instance.

You see the collection that is made here? https://github.com/OpenPerpetuum/PerpetuumServer/blob/Development/src/Perpetuum/Services/EventServices/EventProcessors/EnvironmentalEffectHandler.cs#L37-L45
Instead of having that key on `EffectType` have it key on your Tuple of Time/weather states!
That way you save creating another static dictionary, avoid using `dynamic`, and skip a lookup (O(1), but still more to maintain).
Then `GetEffect` would have a signature like: `ZoneEffect GetEffect(GameTimeState, WeatherState)` or the tuple, either way.  (maybe cleaner to create the tuple in the helper method?)

Sorry I did not look closely enough to realize you were replacing that dictionary, and it works and cleans up a lot of code, but if you want to this extra tweak will really make this a killer PR!
This is the danger of not pulling the code first before reviewing 🤦 

Let me know if this makes sense or if you have any questions.

Reply:
Thank you for the feedback. I think I understand what you mean.
I've pushed [#3976274](https://github.com/OpenPerpetuum/PerpetuumServer/pull/278/commits/397627413f2b45d50132459d2deb5f68d4b33ba5) to try and meet those requirements.

I'm at fault here, for not having better understanding of the code base I'm making changes to.
But I'm happy that I can receive guidance from you, and please be critical. I want to continue learning from this, and hopefully contribute value in the process.

Reply:
Great to hear! I want open source projects like this to be opportunities for people to learn and work on things they otherwise might not be able to build themselves.  But also don't want to make it too hard to contribute either.  Either way I'll work with you to make sure you get your code in the repo.

The changes look good! I'll get you a review and maybe a merge later this evening.
And don't worry, we are still learning the codebase ourselves too ;)